### DISCUSSION TITLE
Question: How to intregrate a Mqtt Client into Bleak

### DISCUSSION THREAD

**Discussion by GitCodeMischer** [hbldh/bleak#671] [Q&A]
* bleak version: 0.12.1
* Python version: 3.9
* Operating System: Windows 10 (Build 19042)

### Description

Hello, I have a question about how do I solve to integrate a Mqtt client make work in combination with the Bleak Library

### What I Did

I try to connect it like this

![image](https://user-images.githubusercontent.com/46819879/137591143-26797f85-2729-4130-8312-e9493a318276.png)

when a message came from the mqtt client i try to connect every time with the Bleak Client async (that takes 4-7 seconds). I know that is not the right way but i dont`t know to stay connect once (global) and accessing it without reconnecting every time. 

_P.S i hope its correct to post Mqtt Client stuff here in a Bleak Issue._

**Preciate your help or Solution**

**Comment by dlech** [comment]
Which MQTT library are you using?

**Reply by GitCodeMischer** [reply]
I am using the  Paho MQTT Client from the Eclipse Foundation

**Reply by dlech** [reply]
Can you share the entire program?

(format it like this)

````
```python
# program here
```
````

**Reply by GitCodeMischer** [reply]
```python
# python3.6
import asyncio
import json
import random
import uuid

from paho.mqtt import client as mqtt_client

from bleak import BleakClient

broker = 'brokername'
port = 1883
client_id = f'python-mqtt-{random.randint(0, 100)}'

def connect_mqtt() -> mqtt_client:
    def on_connect(client, userdata, flags, rc):
        if rc == 0:
            print("Connected to MQTT Broker!",broker)
        else:
            print("Failed to connect, return code %d\n", rc)

    client = mqtt_client.Client(client_id)
    client.on_connect = on_connect
    client.connect(broker, port)
    return client


def subscribe(client: mqtt_client):
    def on_message(client, userdata, msg):
        print(f"Received `{msg.payload.decode()}` from `{msg.topic}` topic")
        asyncio.run(WriteMessageToDevice(msg))

    client.subscribe('/bleak_topic')
    client.on_message = on_message

async def ConnectToDevice():
    async with BleakClient("BE:96:70:00:0B:D9") as client:
        return client

async def WriteMessageToDevice(msg):
    decoded_msg = json.loads(msg.payload.decode())

    async with BleakClient("BE:96:70:00:0B:D9") as client:
        await client.write_gatt_char(uuid.UUID('0000fff3-0000-1000-8000-00805f9b34fb'),bytearray([126, 0, 5, 3, 22, 34, 136, 0, 239],),response=False)

def run():
    client = connect_mqtt()
    subscribe(client)
    #ble_client = asyncio.run(ConnectToDevice())
    #global bleak_client = ble_client
    client.loop_forever()

if __name__ == '__main__':
    run()
```

P.S the connectToDevice Method should work as a global Bleak connection but is not working as expected by the way

**Reply by dlech** [reply]
> ```python
> 
> async def ConnectToDevice():
>     async with BleakClient("BE:96:70:00:0B:D9") as client:
>         return client
> ```

Since `BleakClient` is used with `async with`, it will be disconnected when the method returns. 


> ```python
>     #ble_client = asyncio.run(ConnectToDevice())
> ```

The asyncio run loop needs to be running the entire time for Bleak to work, not just when connecting.

> ```python
> 
>     #global bleak_client = ble_client
> ```

Instead of a global, you could pass the client as an argument to other functions.

--- 

Python 3.9 introduced a [asyncio.to_thread()](https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread) function that can be used to call blocking MQTT methods if there isn't an asyncio version of the MQTT library. 

There is also a similar [loop.run_in_executor()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) method.

**Reply by GitCodeMischer** [reply]
Thanks for your help @dlech
